util/movement-types/src/algs/grouping_heuristic/mod.rs

```diff
@@ -0,0 +1,458 @@
+        }
+    }
+
+    /// Returns true if the outcome is a failure.
+    pub fn is_failure(&self) -> bool {
+        match self {
+            ElementalOutcome::Apply(_) => false,
+            ElementalOutcome::Success => false,
+            ElementalOutcome::Failure(_) => true
+        }
+    }
+
+    /// Returns true if the outcome is an apply.
+    pub fn is_apply(&self) -> bool {
+        match self {
+            ElementalOutcome::Apply(_) => true,
+            ElementalOutcome::Success => false,
+            ElementalOutcome::Failure(_) => false
+        }
+    }
+
+    /// Returns true if the outcome is done.
+    pub fn is_done(&self) -> bool {
+        match self {
+            ElementalOutcome::Apply(_) => false,
+            ElementalOutcome::Success => true,
+            ElementalOutcome::Failure(f) => f.is_terminal()
+        }
+    }
+
+    /// Converts the outcome to a terminal failure.
+    /// Success is not converted to a terminal failure.
+    pub fn to_terminal(self) -> Self {
+        match self {
+            ElementalOutcome::Apply(t) => ElementalOutcome::Failure(ElementalFailure::Terminal(t)),
+            ElementalOutcome::Success => ElementalOutcome::Success,
+            ElementalOutcome::Failure(f) => ElementalOutcome::Failure(f.to_terminal())
+        }
+    }
+
+    /// Converts the outcome to an instrumental failure.
+    /// Success is not converted to an instrumental failure.
+    pub fn to_instrumental(self) -> Self {
+        match self {
+            ElementalOutcome::Apply(t) => ElementalOutcome::Failure(ElementalFailure::Instrumental(t)),
+            ElementalOutcome::Success => ElementalOutcome::Success,
+            ElementalOutcome::Failure(f) => ElementalOutcome::Failure(f.to_instrumental())
+        }
+    }
+
+    /// Converts an outcome to a failure, preserving it's instrumental or terminal status.
+    pub fn to_failures_prefer_instrumental(self) -> Self {
+
+        match self {
+            ElementalOutcome::Apply(t) => ElementalOutcome::Failure(ElementalFailure::Instrumental(t)),
+            ElementalOutcome::Success => ElementalOutcome::Success,
+            ElementalOutcome::Failure(f) => ElementalOutcome::Failure(f)
+        }
+
+    }
+
+    /// Converts the outcome to an apply outcome.
+    /// Success is not converted to an apply outcome.
+    pub fn to_apply(self) -> Self {
+        match self {
+            ElementalOutcome::Apply(t) => ElementalOutcome::Apply(t),
+            ElementalOutcome::Success => ElementalOutcome::Success,
+            ElementalOutcome::Failure(f) => ElementalOutcome::Apply(f.into_inner())
+        }
+    }
+
+}
+
+/// The outcomes for a particular group in a grouping heuristic.
+pub struct GroupingOutcome<T>(pub Vec<ElementalOutcome<T>>);
+
+impl <T> PartialEq for GroupingOutcome<T> where T: PartialEq {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.iter().zip(other.0.iter()).all(|(a, b)| a == b)
+    }
+}
```

Comment:
The impl you'd get with `#[derive(PartialEq)]` does not only get same effect with no code, it's faster because it delegates to slice comparison which is optimized and specialized for bitwise-comparable values, which you can't do outside of the standard library.